File: src/lib/mobile/geofence.ts
// ═══════════════════════════════════════════════════════════════
// GEOFENCE — location validation for attendance
// ═══════════════════════════════════════════════════════════════
// Pure, so it tests without a device. Shared by the mobile app and the server:
// the phone checks so it can tell the user *before* they try, and the server
// checks because a phone is an untrusted client.
//
// The thing to understand about this module is that it is adversarial. Clock-in
// location decides pay. There are published apps whose only purpose is to feed
// a fake location to attendance software, and they work by setting the mock
// provider flag that Android exposes and iOS does not. So:
//
//   - Accuracy is treated as a radius, not a decoration. A fix accurate to
//     500 m "inside" a 100 m fence tells you nothing.
//   - A suspiciously perfect fix is more concerning than a rough one. Real GPS
//     is never accurate to 1 m outdoors on a phone.
//   - Implausible movement between two fixes is flagged, because that is what
//     a spoofed jump looks like and a genuine one cannot.
//
// None of these are certainties. They are reasons for a human to look, which
// is why the result carries a reason string rather than a verdict.

export interface Coordinates {
  latitude: number;
  longitude: number;
  /** Reported accuracy radius in metres. */
  accuracyMetres?: number;
  /** Epoch milliseconds. */
  capturedAt?: number;
  /** Android exposes this; iOS does not, so absence proves nothing. */
  isMocked?: boolean;
}

export interface Geofence {
  id: string;
  name: string;
  latitude: number;
  longitude: number;
  radiusMetres: number;
}

const EARTH_RADIUS_METRES = 6_371_008.8;

/**
 * Great-circle distance between two points, in metres.
 *
 * Haversine rather than the equirectangular approximation: the approximation
 * is faster and wrong by enough to matter at the edge of a 50 m office fence,
 * which is exactly where the answer is contested.
 */
export function distanceMetres(a: Coordinates, b: Coordinates): number {
  assertCoordinates(a);
  assertCoordinates(b);

  const lat1 = toRadians(a.latitude);
  const lat2 = toRadians(b.latitude);
  const deltaLat = toRadians(b.latitude - a.latitude);
  const deltaLon = toRadians(b.longitude - a.longitude);

  const h =
    Math.sin(deltaLat / 2) ** 2 +
    Math.cos(lat1) * Math.cos(lat2) * Math.sin(deltaLon / 2) ** 2;

  return 2 * EARTH_RADIUS_METRES * Math.asin(Math.min(1, Math.sqrt(h)));
}

function toRadians(degrees: number): number {
  return (degrees * Math.PI) / 180;
}

function assertCoordinates(point: Coordinates): void {
  if (!Number.isFinite(point.latitude) || !Number.isFinite(point.longitude)) {
    throw new Error("Coordinates must be finite numbers");
  }
  if (point.latitude < -90 || point.latitude > 90) {
    throw new Error("Latitude must be between -90 and 90");
  }
  if (point.longitude < -180 || point.longitude > 180) {
    throw new Error("Longitude must be between -180 and 180");
  }
}

export type FenceConfidence = "inside" | "probably_inside" | "uncertain" | "outside";

export interface FenceResult {
  fence: Geofence | null;
  distanceMetres: number | null;
  confidence: FenceConfidence;
  /** Shown to the user, so it says what to do rather than what went wrong. */
  message: string;
}

/**
 * Which fence a position falls in, accounting for accuracy.
 *
 * Accuracy is a radius of uncertainty, so there are three cases, not two:
 * definitely inside, definitely outside, and a fix too imprecise to say. The
 * third is common indoors and near tall buildings, and collapsing it into
 * either of the others is how an office worker gets marked absent for standing
 * next to a window.
 */
export function locateWithin(
  position: Coordinates,
  fences: Geofence[],
  options: { maxAccuracyMetres?: number } = {}
): FenceResult {
  if (fences.length === 0) {
    return {
      fence: null,
      distanceMetres: null,
      confidence: "uncertain",
      message: "No work locations are configured",
    };
  }

  const accuracy = position.accuracyMetres ?? 0;
  const maxAccuracy = options.maxAccuracyMetres ?? 200;

  // A fix this rough cannot place someone inside anything smaller than a
  // postcode, and pretending otherwise is worse than asking them to move.
  if (accuracy > maxAccuracy) {
    return {
      fence: null,
      distanceMetres: null,
      confidence: "uncertain",
      message: `Your location is only accurate to about ${Math.round(accuracy)} m. Step outside or near a window and try again.`,
    };
  }

  const ranked = fences
    .map((fence) => ({ fence, distance: distanceMetres(position, fence) }))
    .sort((a, b) => a.distance - b.distance);

  const nearest = ranked[0];

  // Unreachable — `fences.length === 0` returned above — but stated rather
  // than assumed. Under noUncheckedIndexedAccess an array index is possibly
  // undefined, and the alternative is a non-null assertion that would keep
  // being correct only for as long as the guard above stays where it is.
  if (!nearest) {
    return {
      fence: null,
      distanceMetres: null,
      confidence: "uncertain",
      message: "No work locations are configured",
    };
  }

  // Certainly inside: even the far edge of the uncertainty circle is within
  // the fence.
  if (nearest.distance + accuracy <= nearest.fence.radiusMetres) {
    return {
      fence: nearest.fence,
      distanceMetres: Math.round(nearest.distance),
      confidence: "inside",
      message: `At ${nearest.fence.name}`,
    };
  }

  // Certainly outside: even the near edge is beyond the fence.
  if (nearest.distance - accuracy > nearest.fence.radiusMetres) {
    return {
      fence: null,
      distanceMetres: Math.round(nearest.distance),
      confidence: "outside",
      message: `You are about ${formatDistance(nearest.distance - nearest.fence.radiusMetres)} from ${nearest.fence.name}`,
    };
  }

  // The circle straddles the boundary. The centre reading decides which way to
  // lean, but the caller is told it is a lean.
  return {
    fence: nearest.distance <= nearest.fence.radiusMetres ? nearest.fence : null,
    distanceMetres: Math.round(nearest.distance),
    confidence: nearest.distance <= nearest.fence.radiusMetres ? "probably_inside" : "uncertain",
    message: `You seem to be at the edge of ${nearest.fence.name}. Move a little closer if this is refused.`,
  };
}

function formatDistance(metres: number): string {
  return metres >= 1000 ? `${(metres / 1000).toFixed(1)} km` : `${Math.round(metres)} m`;
}

// ─── Spoofing signals ────────────────────────────────────────

export interface SpoofSignal {
  code: "mock_provider" | "impossible_speed" | "implausible_accuracy" | "stale_fix";
  severity: "high" | "medium";
  detail: string;
}

/**
 * Reasons to distrust a position.
 *
 * Signals, not a verdict. Every one of these has an innocent explanation —
 * a developer with mock locations left on, a passenger on a train, a fix
 * cached while the app was backgrounded. The output is for a human reviewing
 * an exception report, and treating any of it as proof would deny someone
 * their pay on a heuristic.
 */
export function spoofSignals(
  position: Coordinates,
  previous?: Coordinates,
  now = Date.now()
): SpoofSignal[] {
  const signals: SpoofSignal[] = [];

  if (position.isMocked) {
    // The only signal the operating system asserts rather than something we
    // inferred, so it is the only high-severity one on its own.
    signals.push({
      code: "mock_provider",
      severity: "high",
      detail: "The device reported this location came from a mock provider",
    });
  }

  if (position.accuracyMetres !== undefined && position.accuracyMetres > 0 && position.accuracyMetres < 3) {
    // Consumer GPS does not achieve sub-3 m outdoors. A fabricated fix often
    // reports an implausibly confident accuracy because the faker picked a
    // number that sounded good.
    signals.push({
      code: "implausible_accuracy",
      severity: "medium",
      detail: `Reported accuracy of ${position.accuracyMetres} m is better than consumer GPS achieves`,
    });
  }

  if (position.capturedAt !== undefined) {
    const ageMs = now - position.capturedAt;
    // Five minutes. A fix older than that was taken somewhere the user may no
    // longer be.
    if (ageMs > 5 * 60_000) {
      signals.push({
        code: "stale_fix",
        severity: "medium",
        detail: `This location was captured ${Math.round(ageMs / 60_000)} minutes ago`,
      });
    }
  }

  // Compared against undefined rather than tested for truthiness: a timestamp
  // of 0 is a real value, and `if (previous?.capturedAt)` skips the whole
  // check for it. The same shape of bug as treating Number("") as absent.
  if (previous?.capturedAt !== undefined && position.capturedAt !== undefined) {
    const seconds = (position.capturedAt - previous.capturedAt) / 1000;

    if (seconds > 0) {
      const metres = distanceMetres(previous, position);
      const kmh = (metres / seconds) * 3.6;

      // 900 km/h is airliner cruise. Below that, "impossible" would flag
      // anyone on a flight, and people do clock in from airports.
      if (kmh > 900) {
        signals.push({
          code: "impossible_speed",
          severity: "high",
          detail: `${Math.round(metres)} m in ${Math.round(seconds)} s implies ${Math.round(kmh)} km/h`,
        });
      }
    }
  }

  return signals;
}

export interface ClockInVerdict {
  allowed: boolean;
  requiresReview: boolean;
  fence: Geofence | null;
  confidence: FenceConfidence;
  signals: SpoofSignal[];
  message: string;
}

/**
 * Whether a clock-in should be accepted.
 *
 * `requiresReview` rather than a refusal on a spoofing signal. Someone whose
 * phone reports a mock provider because a developer tool is installed still
 * turned up for work, and refusing them costs a day's pay on a guess. Flagging
 * it puts the decision in front of a person who can ask.
 *
 * An uncertain fix is allowed and flagged for the same reason: indoor GPS is
 * genuinely poor, and the alternative is that everyone in the middle of a
 * building is marked absent.
 */
export function evaluateClockIn(
  position: Coordinates,
  fences: Geofence[],
  options: {
    previous?: Coordinates;
    now?: number;
    maxAccuracyMetres?: number;
    /** Allows clocking in from anywhere, for field and remote staff. */
    allowAnywhere?: boolean;
  } = {}
): ClockInVerdict {
  const signals = spoofSignals(position, options.previous, options.now);

  if (options.allowAnywhere) {
    return {
      allowed: true,
      requiresReview: signals.some((s) => s.severity === "high"),
      fence: null,
      confidence: "inside",
      signals,
      message: "Clocked in",
    };
  }

  const located = locateWithin(position, fences, {
    maxAccuracyMetres: options.maxAccuracyMetres,
  });

  if (located.confidence === "outside") {
    return {
      allowed: false,
      requiresReview: false,
      fence: null,
      confidence: "outside",
      signals,
      message: located.message,
    };
  }

  return {
    allowed: true,
    requiresReview:
      signals.some((s) => s.severity === "high") ||
      located.confidence !== "inside",
    fence: located.fence,
    confidence: located.confidence,
    signals,
    message: located.message,
  };
}

File: android/app/src/main/java/com/circuvent/hrms/domain/Geofence.kt
package com.circuvent.hrms.domain

import kotlin.math.abs
import kotlin.math.asin
import kotlin.math.cos
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.roundToLong
import kotlin.math.sin
import kotlin.math.sqrt

/**
 * GEOFENCE — is this person at work?
 *
 * The single most consequential piece of arithmetic in the app: it decides
 * whether somebody's clock-in is accepted, which decides whether they are paid
 * for the day.
 *
 * The Earth radius is 6,371,008.8 m and that figure is load-bearing. The
 * previous generation of this product briefly had two haversine
 * implementations — the server used 6,371,000 and the phone used 6,371,008.8 —
 * so the two disagreed by about a metre per kilometre. Small, until it puts
 * somebody on opposite sides of a 50 m office boundary depending on which one
 * you asked, and the answer decides whether they were at work.
 *
 * The server runs the same calculation. This copy exists so a refusal is
 * instant and explains itself, rather than arriving as a 403 after a round
 * trip. It is not security: the phone is not trusted and the server checks
 * again.
 */
object Geofence {

    /** IUGG mean Earth radius. Must match the server. */
    const val EARTH_RADIUS_METRES = 6_371_008.8

    data class Coordinates(
        val latitude: Double,
        val longitude: Double,
        val accuracyMetres: Double? = null,
        /** Epoch milliseconds of the fix. */
        val capturedAt: Long? = null,
        /** The OS says this came from a mock provider. */
        val isMocked: Boolean = false,
    )

    data class Fence(
        val id: String,
        val name: String,
        val latitude: Double,
        val longitude: Double,
        val radiusMetres: Double,
    )

    enum class Confidence { INSIDE, PROBABLY_INSIDE, UNCERTAIN, OUTSIDE }

    data class FenceResult(
        val fence: Fence?,
        val distanceMetres: Int?,
        val confidence: Confidence,
        val message: String,
    )

    enum class Severity { HIGH, MEDIUM }

    data class Signal(val code: String, val severity: Severity, val detail: String)

    data class Verdict(
        val allowed: Boolean,
        val requiresReview: Boolean,
        val fence: Fence?,
        val confidence: Confidence,
        val signals: List<Signal>,
        val message: String,
    )

    /** Accuracy beyond which a fix cannot decide a 50 m boundary. */
    const val DEFAULT_MAX_ACCURACY_METRES = 100.0

    private fun requireValid(point: Coordinates) {
        require(point.latitude.isFinite() && point.longitude.isFinite()) {
            "Coordinates must be finite numbers"
        }
        require(point.latitude in -90.0..90.0) { "Latitude must be between -90 and 90" }
        require(point.longitude in -180.0..180.0) { "Longitude must be between -180 and 180" }
    }

    /**
     * Great-circle distance between two points, in metres.
     *
     * Haversine rather than the equirectangular approximation: the
     * approximation is faster and wrong by enough to matter at the edge of a
     * 50 m office fence, which is exactly where the answer is contested.
     */
    fun distanceMetres(a: Coordinates, b: Coordinates): Double {
        requireValid(a)
        requireValid(b)

        val lat1 = Math.toRadians(a.latitude)
        val lat2 = Math.toRadians(b.latitude)
        val deltaLat = Math.toRadians(b.latitude - a.latitude)
        val deltaLon = Math.toRadians(b.longitude - a.longitude)

        val h = sin(deltaLat / 2) * sin(deltaLat / 2) +
            cos(lat1) * cos(lat2) * sin(deltaLon / 2) * sin(deltaLon / 2)

        // min(1.0, …) guards the floating-point case where h creeps just over
        // 1 for antipodal points and asin returns NaN.
        return 2 * EARTH_RADIUS_METRES * asin(min(1.0, sqrt(h)))
    }

    private fun distanceTo(position: Coordinates, fence: Fence): Double =
        distanceMetres(
            position,
            Coordinates(latitude = fence.latitude, longitude = fence.longitude),
        )

    private fun formatDistance(metres: Double): String =
        if (metres >= 1000) "%.1f km".format(metres / 1000) else "${metres.roundToInt()} m"

    /**
     * Which fence, if any, this position is within — and how sure we are.
     *
     * The accuracy radius is treated as a real uncertainty rather than
     * ignored. A fix accurate to 80 m cannot decide a 50 m boundary, and
     * pretending otherwise is worse than asking the person to move.
     */
    fun locateWithin(
        position: Coordinates,
        fences: List<Fence>,
        maxAccuracyMetres: Double = DEFAULT_MAX_ACCURACY_METRES,
    ): FenceResult {
        if (fences.isEmpty()) {
            return FenceResult(null, null, Confidence.UNCERTAIN, "No work locations are configured")
        }

        val accuracy = position.accuracyMetres ?: 0.0
        if (accuracy > maxAccuracyMetres) {
            return FenceResult(
                fence = null,
                distanceMetres = null,
                confidence = Confidence.UNCERTAIN,
                message = "Your location is only accurate to about ${accuracy.roundToInt()} m. " +
                    "Step outside or near a window and try again.",
            )
        }

        // minByOrNull rather than sorting and indexing. The TypeScript version
        // needed a comment and a dead branch to explain why an index into a
        // non-empty array was still possibly undefined under its own compiler
        // settings; Kotlin's nullable return says it once.
        val nearest = fences.minByOrNull { distanceTo(position, it) }
            ?: return FenceResult(null, null, Confidence.UNCERTAIN, "No work locations are configured")

        val distance = distanceTo(position, nearest)

        // Certainly inside: even the far edge of the uncertainty circle is
        // within the fence.
        if (distance + accuracy <= nearest.radiusMetres) {
            return FenceResult(nearest, distance.roundToInt(), Confidence.INSIDE, "At ${nearest.name}")
        }

        // Certainly outside: even the near edge is beyond it.
        if (distance - accuracy > nearest.radiusMetres) {
            return FenceResult(
                fence = null,
                distanceMetres = distance.roundToInt(),
                confidence = Confidence.OUTSIDE,
                message = "You are about ${formatDistance(distance - nearest.radiusMetres)} from ${nearest.name}",
            )
        }

        // The circle straddles the boundary. The centre reading decides which
        // way to lean, but the caller is told that it is a lean.
        val leansInside = distance <= nearest.radiusMetres
        return FenceResult(
            fence = if (leansInside) nearest else null,
            distanceMetres = distance.roundToInt(),
            confidence = if (leansInside) Confidence.PROBABLY_INSIDE else Confidence.UNCERTAIN,
            message = "You seem to be at the edge of ${nearest.name}. Move a little closer if this is refused.",
        )
    }

    /**
     * Reasons to distrust a position.
     *
     * Signals, not a verdict. Every one of these has an innocent explanation —
     * a developer with mock locations left on, a passenger on a train, a fix
     * cached while the app was backgrounded. The output is for a human
     * reviewing an exception report, and treating any of it as proof would
     * deny somebody their pay on a heuristic.
     */
    fun spoofSignals(
        position: Coordinates,
        previous: Coordinates? = null,
        now: Long = System.currentTimeMillis(),
    ): List<Signal> {
        val signals = mutableListOf<Signal>()

        if (position.isMocked) {
            // The only signal the operating system asserts rather than
            // something we inferred, so it is the only high-severity one on
            // its own.
            signals += Signal(
                "mock_provider",
                Severity.HIGH,
                "The device reports this location came from a mock provider",
            )
        }

        val accuracy = position.accuracyMetres
        if (accuracy != null && accuracy in 0.0..1.0) {
            // Consumer GNSS does not do sub-metre. A reading this good is more
            // likely a fabricated one than a real fix.
            signals += Signal(
                "implausible_accuracy",
                Severity.MEDIUM,
                "An accuracy of ${"%.1f".format(accuracy)} m is better than consumer hardware achieves",
            )
        }

        val capturedAt = position.capturedAt
        // Compared against null, not falsiness. A timestamp of 0 is falsy, and
        // in the previous implementation that skipped the check entirely — the
        // same shape as treating an empty string as an absent number.
        if (capturedAt != null) {
            val age = now - capturedAt
            if (age > STALE_FIX_MS) {
                signals += Signal(
                    "stale_fix",
                    Severity.MEDIUM,
                    "The fix is ${(age / 60_000)} minutes old",
                )
            }

            if (previous?.capturedAt != null) {
                val elapsed = capturedAt - previous.capturedAt
                if (elapsed > 0) {
                    val metres = distanceMetres(previous, position)
                    val metresPerSecond = metres / (elapsed / 1000.0)
                    if (metresPerSecond > IMPOSSIBLE_SPEED_MPS && metres > 500) {
                        signals += Signal(
                            "impossible_speed",
                            Severity.MEDIUM,
                            "That is ${(metresPerSecond * 3.6).roundToLong()} km/h since the last fix",
                        )
                    }
                }
            }
        }

        return signals
    }

    /** Older than this and the fix may predate the tap that asked for it. */
    private const val STALE_FIX_MS = 5 * 60 * 1000L

    /** Faster than a commercial aircraft; 300 m/s is about 1,080 km/h. */
    private const val IMPOSSIBLE_SPEED_MPS = 300.0

    /**
     * The decision.
     *
     * `allowAnywhere` is for remote and field staff, whose employer has not
     * configured a fence for them. They still get their signals recorded, so a
     * mock provider is visible on the exception report — it simply does not
     * block the punch.
     *
     * Nothing here refuses a punch on a spoofing signal alone. A signal is a
     * reason for a human to look, and an automatic refusal on a heuristic is
     * how somebody loses a day's pay to a train journey.
     */
    fun evaluateClockIn(
        position: Coordinates,
        fences: List<Fence>,
        previous: Coordinates? = null,
        now: Long = System.currentTimeMillis(),
        maxAccuracyMetres: Double = DEFAULT_MAX_ACCURACY_METRES,
        allowAnywhere: Boolean = false,
    ): Verdict {
        val signals = spoofSignals(position, previous, now)
        val highSeverity = signals.any { it.severity == Severity.HIGH }

        if (allowAnywhere) {
            return Verdict(
                allowed = true,
                requiresReview = highSeverity,
                fence = null,
                confidence = Confidence.INSIDE,
                signals = signals,
                message = "Clocked in",
            )
        }

        val located = locateWithin(position, fences, maxAccuracyMetres)

        if (located.confidence == Confidence.OUTSIDE) {
            return Verdict(
                allowed = false,
                requiresReview = false,
                fence = null,
                confidence = Confidence.OUTSIDE,
                signals = signals,
                message = located.message,
            )
        }

        return Verdict(
            allowed = true,
            requiresReview = highSeverity || located.confidence != Confidence.INSIDE,
            fence = located.fence,
            confidence = located.confidence,
            signals = signals,
            message = located.message,
        )
    }

    /** Kept for the tests that assert the two implementations agree. */
    internal fun absDifference(a: Double, b: Double): Double = abs(a - b)
}
